server/src/main/java/org/elasticsearch/health/components/controller/ClusterCoordination.java

```diff
@@ -60,8 +65,32 @@ public static HealthComponentResult createClusterCoordinationComponent(
             }
         );
 
-        // Only a single indicator currently so it determines the status
-        final HealthStatus status = instanceHasMaster.status();
-        return new HealthComponentResult(NAME, status, Collections.singletonMap(INSTANCE_HAS_MASTER_NAME, instanceHasMaster));
+        var noEligibleMasterNodes = noEligibleMasterNodes(clusterState);
+        var lackOfQuorum = lackOfQuorum(clusterState);
+        final HealthStatus status = HealthStatus.merge(
+            Stream.of(instanceHasMaster, noEligibleMasterNodes, lackOfQuorum).map(HealthIndicatorResult::status)
+        );
+        return new HealthComponentResult(
+            NAME,
+            status,
+            Map.of(INSTANCE_HAS_MASTER_NAME, instanceHasMaster, NO_ELIGIBLE_MASTERS, noEligibleMasterNodes, LACK_OF_QUORUM, lackOfQuorum)
+        );
+    }
+
+    static HealthIndicatorResult noEligibleMasterNodes(ClusterState clusterState) {
+        for (DiscoveryNode node : clusterState.nodes()) {
+            if (node.getRoles().contains(DiscoveryNodeRole.MASTER_ROLE)) {
+                return HealthIndicatorResult.of(NO_ELIGIBLE_MASTERS, NAME, HealthStatus.GREEN, "There are eligible master nodes");
+            }
+        }
+        return HealthIndicatorResult.of(NO_ELIGIBLE_MASTERS, NAME, HealthStatus.RED, "No eligible master nodes");
+    }
+
+    static HealthIndicatorResult lackOfQuorum(ClusterState clusterState) {
+        boolean hasQuorum = clusterState.getLastCommittedConfiguration()
+            .hasQuorum(clusterState.getNodes().getNodes().values().stream().map(DiscoveryNode::getId).collect(Collectors.toList()));
+        return hasQuorum
+            ? HealthIndicatorResult.of(LACK_OF_QUORUM, NAME, HealthStatus.GREEN, "Instance can form a quorum")
+            : HealthIndicatorResult.of(LACK_OF_QUORUM, NAME, HealthStatus.RED, "Not enough nodes with voting rights to form a quorum");
     }
```

Comment:
I still have a concern that these indicators don't go `RED` again once a node has joined a cluster. I don't see a test for this.

I also don't think we can detect whether there are enough nodes to form a quorum by looking only at local state, so I think that a local-only check for a quorum of nodes will give false information (both false positives and false negatives).

Reply:
I thought that adding new nodes shouldn't affect th `hasQuorum` calculation? I tried to come up with a simple test for that, not sure if it's correct:
```java
@ESIntegTestCase.ClusterScope(numDataNodes = 0, scope = ESIntegTestCase.Scope.TEST)
public class GetLackOfQourumIT extends ESIntegTestCase {

    @Override
    protected Collection<Class<? extends Plugin>> nodePlugins() {
        return Collections.singletonList(MockTransportService.TestPlugin.class);
    }

    public void testNotWaitForQuorumCopies() throws Exception {
        logger.info("--> starting 3 nodes");
        internalCluster().startMasterOnlyNodes(3);
        ensureStableCluster(3);

        System.out.println("Health Indicator Result: " + lackOfQourumIndicator());

        internalCluster().startDataOnlyNode();
        ensureStableCluster(4);

        System.out.println("Health Indicator Result: " + lackOfQourumIndicator());
    }

    private HealthIndicatorResult lackOfQourumIndicator() throws InterruptedException, ExecutionException {
        return client().execute(GetHealthAction.INSTANCE, new GetHealthAction.Request())
            .get()
            .getComponents()
            .stream()
            .filter(c -> c.name().equals("cluster_coordination"))
            .findAny()
            .orElseThrow()
            .indicators()
            .get("lack-of-quorum");
    }
}
```

Returns:
```
[2022-02-08T04:30:24,186][INFO ][o.e.h.c.c.ClusterCoordination] [testNotWaitForQuorumCopies] Checking lack of quorum
[2022-02-08T04:30:24,187][INFO ][o.e.h.c.c.ClusterCoordination] [testNotWaitForQuorumCopies] Cluster state nodes [HX98_TapRFeZ_AHlcV2aKQ, qjNaWBlJSASRk9SVoPXWRA, fBNx2MtsRXK5wGxV3mTiQw]
[2022-02-08T04:30:24,187][INFO ][o.e.h.c.c.ClusterCoordination] [testNotWaitForQuorumCopies] Voting configuration [HX98_TapRFeZ_AHlcV2aKQ, qjNaWBlJSASRk9SVoPXWRA, fBNx2MtsRXK5wGxV3mTiQw]
Health Indicator Result: HealthIndicatorResult[name=lack-of-quorum, component=cluster_coordination, status=GREEN, summary=Instance can form a quorum, details=HealthIndicatorDetails{}]
...

[2022-02-08T04:30:24,384][INFO ][o.e.h.c.c.ClusterCoordination] [testNotWaitForQuorumCopies] Checking lack of quorum
[2022-02-08T04:30:24,384][INFO ][o.e.h.c.c.ClusterCoordination] [testNotWaitForQuorumCopies] Cluster state nodes [fBNx2MtsRXK5wGxV3mTiQw, HX98_TapRFeZ_AHlcV2aKQ, AoIBBsomQWOU1lLjFlx3tQ, qjNaWBlJSASRk9SVoPXWRA]
[2022-02-08T04:30:24,384][INFO ][o.e.h.c.c.ClusterCoordination] [testNotWaitForQuorumCopies] Voting configuration [qjNaWBlJSASRk9SVoPXWRA, HX98_TapRFeZ_AHlcV2aKQ, fBNx2MtsRXK5wGxV3mTiQw]
```

Reply:
> I also don't think we can detect whether there are enough nodes to form a quorum by looking only at local state, so I think that a local-only check for a quorum of nodes will give false information (both false positives and false negatives).

When we discussed the indicators for the 1st phase, we (#es-health-api) agreed to rely only on the cluster state for the calculation of indicators. What would you suggest a proper check for the lack of quorum should look like?

Reply:
The proper calculation is covered in the decision tree I shared in Slack - it requires collecting data from all the master nodes. The information needed is recorded in the logs so today we have to collect logs from all the nodes and correlate them. It would be awfully nice if ES did that itself, but I also think it would be fine if the first draft of this API didn't report anything about quorums at all. This would seem preferable to reporting something incorrect.

Reply:
Here's an example of a way to get a cluster into a state where different nodes have different opinions about whether there's a quorum or not:

<details>

<summary>Expand for code</summary>

```java
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

package org.elasticsearch.health;

import org.elasticsearch.cluster.node.DiscoveryNode;
import org.elasticsearch.common.util.CollectionUtils;
import org.elasticsearch.plugins.Plugin;
import org.elasticsearch.test.ESIntegTestCase;
import org.elasticsearch.test.transport.MockTransportService;
import org.elasticsearch.transport.Transport;
import org.elasticsearch.transport.TransportService;

import java.util.Collection;
import java.util.Set;

import static org.hamcrest.Matchers.anyOf;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.hasSize;
import static org.hamcrest.Matchers.not;
import static org.hamcrest.Matchers.nullValue;

@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0, autoManageMasterNodes = false)
public class QuorumMismatchIT extends ESIntegTestCase {

    @Override
    protected Collection<Class<? extends Plugin>> nodePlugins() {
        return CollectionUtils.appendToCopy(super.nodePlugins(), MockTransportService.TestPlugin.class);
    }

    public void testQuorumMismatch() throws Exception {
        internalCluster().setBootstrapMasterNodeIndex(0);
        internalCluster().startMasterOnlyNodes(4);
        ensureGreen();

        final Set<String> originalVotingNodes = client().admin()
            .cluster()
            .prepareState()
            .get()
            .getState()
            .getLastAcceptedConfiguration()
            .getNodeIds();
        assertThat(originalVotingNodes, hasSize(3));
        final DiscoveryNode nonVotingNode = client().admin()
            .cluster()
            .prepareState()
            .get()
            .getState()
            .nodes()
            .stream()
            .filter(n -> originalVotingNodes.contains(n.getId()) == false)
            .findAny()
            .orElseThrow(() -> new AssertionError("nonvoting node"));

        final String originalMasterName = internalCluster().getMasterName();
        final MockTransportService masterTransportService = (MockTransportService) internalCluster().getInstance(
            TransportService.class,
            originalMasterName
        );

        masterTransportService.addSendBehavior((connection, requestId, action, request, options) -> {
            connection.close();
            connection.sendRequest(requestId, action, request, options);
        });

        ensureStableCluster(3, nonVotingNode.getName());

        final String newMasterName = internalCluster().getMasterName();
        assertThat(newMasterName, not(anyOf(nullValue(), equalTo(originalMasterName))));
        assertBusy(
            () -> assertNotEquals(
                originalVotingNodes,
                client(nonVotingNode.getName()).admin()
                    .cluster()
                    .prepareState()
                    .get()
                    .getState()
                    .getLastAcceptedConfiguration()
                    .getNodeIds()
            )
        );

        internalCluster().stopNode(nonVotingNode.getName());
        internalCluster().stopNode(newMasterName);

        masterTransportService.addSendBehavior(Transport.Connection::sendRequest);

        Thread.sleep(25000);

        for (String nodeName : internalCluster().getNodeNames()) {
            internalCluster().stopNode(nodeName);
        }
    }
}
```
</details>

I didn't try and make this reliable, but on most runs you should see logs like this towards the end:

```
[2022-02-07T14:22:15,631][WARN ][o.e.c.c.ClusterFormationFailureHelper] [node_t0] master not discovered or elected yet, an election requires at least 2 nodes with ids from [-2kPcYGYS6-YrXtxtpKLqA, 9Cl_2RR7SjSe7021DLHnXA, HzRJBu3OReW9OoEFx44J_A], have discovered possible quorum [{node_t0}{9Cl_2RR7SjSe7021DLHnXA}{Tab-Ijb3QEu792QtkQXD-g}{127.0.0.1}{127.0.0.1:60944}{m}, {node_t2}{-2kPcYGYS6-YrXtxtpKLqA}{-yP93BdVQ3a3vAwYipvIgQ}{127.0.0.1}{127.0.0.1:60946}{m}]; discovery will continue using [127.0.0.1:60947, 127.0.0.1:60946, 127.0.0.1:60945] from hosts providers and [{node_t2}{-2kPcYGYS6-YrXtxtpKLqA}{-yP93BdVQ3a3vAwYipvIgQ}{127.0.0.1}{127.0.0.1:60946}{m}, {node_t0}{9Cl_2RR7SjSe7021DLHnXA}{Tab-Ijb3QEu792QtkQXD-g}{127.0.0.1}{127.0.0.1:60944}{m}, {node_t3}{m_60Bs56TkGjspK5hFrJPQ}{fAPax2AAQjScHgyIXG_08w}{127.0.0.1}{127.0.0.1:60945}{m}, {node_t1}{HzRJBu3OReW9OoEFx44J_A}{-6HiUyhkRC2hLMEUy2wxKQ}{127.0.0.1}{127.0.0.1:60947}{m}] from last-known cluster state; node term 3, last-accepted version 5 in term 3
[2022-02-07T14:22:18,682][WARN ][o.e.c.c.ClusterFormationFailureHelper] [node_t2] master not discovered or elected yet, an election requires at least 2 nodes with ids from [-2kPcYGYS6-YrXtxtpKLqA, HzRJBu3OReW9OoEFx44J_A, m_60Bs56TkGjspK5hFrJPQ], have only discovered non-quorum [{node_t2}{-2kPcYGYS6-YrXtxtpKLqA}{-yP93BdVQ3a3vAwYipvIgQ}{127.0.0.1}{127.0.0.1:60946}{m}, {node_t0}{9Cl_2RR7SjSe7021DLHnXA}{Tab-Ijb3QEu792QtkQXD-g}{127.0.0.1}{127.0.0.1:60944}{m}]; discovery will continue using [127.0.0.1:60944, 127.0.0.1:60947, 127.0.0.1:60945] from hosts providers and [{node_t2}{-2kPcYGYS6-YrXtxtpKLqA}{-yP93BdVQ3a3vAwYipvIgQ}{127.0.0.1}{127.0.0.1:60946}{m}, {node_t1}{HzRJBu3OReW9OoEFx44J_A}{-6HiUyhkRC2hLMEUy2wxKQ}{127.0.0.1}{127.0.0.1:60947}{m}] from last-known cluster state; node term 3, last-accepted version 8 in term 3
...
[2022-02-07T14:22:25,632][WARN ][o.e.c.c.ClusterFormationFailureHelper] [node_t0] master not discovered or elected yet, an election requires at least 2 nodes with ids from [-2kPcYGYS6-YrXtxtpKLqA, 9Cl_2RR7SjSe7021DLHnXA, HzRJBu3OReW9OoEFx44J_A], have discovered possible quorum [{node_t0}{9Cl_2RR7SjSe7021DLHnXA}{Tab-Ijb3QEu792QtkQXD-g}{127.0.0.1}{127.0.0.1:60944}{m}, {node_t2}{-2kPcYGYS6-YrXtxtpKLqA}{-yP93BdVQ3a3vAwYipvIgQ}{127.0.0.1}{127.0.0.1:60946}{m}]; discovery will continue using [127.0.0.1:60947, 127.0.0.1:60946, 127.0.0.1:60945] from hosts providers and [{node_t2}{-2kPcYGYS6-YrXtxtpKLqA}{-yP93BdVQ3a3vAwYipvIgQ}{127.0.0.1}{127.0.0.1:60946}{m}, {node_t0}{9Cl_2RR7SjSe7021DLHnXA}{Tab-Ijb3QEu792QtkQXD-g}{127.0.0.1}{127.0.0.1:60944}{m}, {node_t3}{m_60Bs56TkGjspK5hFrJPQ}{fAPax2AAQjScHgyIXG_08w}{127.0.0.1}{127.0.0.1:60945}{m}, {node_t1}{HzRJBu3OReW9OoEFx44J_A}{-6HiUyhkRC2hLMEUy2wxKQ}{127.0.0.1}{127.0.0.1:60947}{m}] from last-known cluster state; node term 3, last-accepted version 5 in term 3
[2022-02-07T14:22:28,684][WARN ][o.e.c.c.ClusterFormationFailureHelper] [node_t2] master not discovered or elected yet, an election requires at least 2 nodes with ids from [-2kPcYGYS6-YrXtxtpKLqA, HzRJBu3OReW9OoEFx44J_A, m_60Bs56TkGjspK5hFrJPQ], have only discovered non-quorum [{node_t2}{-2kPcYGYS6-YrXtxtpKLqA}{-yP93BdVQ3a3vAwYipvIgQ}{127.0.0.1}{127.0.0.1:60946}{m}, {node_t0}{9Cl_2RR7SjSe7021DLHnXA}{Tab-Ijb3QEu792QtkQXD-g}{127.0.0.1}{127.0.0.1:60944}{m}]; discovery will continue using [127.0.0.1:60944, 127.0.0.1:60947, 127.0.0.1:60945] from hosts providers and [{node_t2}{-2kPcYGYS6-YrXtxtpKLqA}{-yP93BdVQ3a3vAwYipvIgQ}{127.0.0.1}{127.0.0.1:60946}{m}, {node_t1}{HzRJBu3OReW9OoEFx44J_A}{-6HiUyhkRC2hLMEUy2wxKQ}{127.0.0.1}{127.0.0.1:60947}{m}] from last-known cluster state; node term 3, last-accepted version 8 in term 3
```

Note that `node_t0` reports `have discovered possible quorum` whereas `node_t2` says `have only discovered non-quorum`. In this case `node_t2` is correct, the cluster cannot form because it's lost quorum, but `node_t0` thinks the problem is something else.

Reply:
I think we should obviously go with whatever David thinks is appropriate here. The first phase should deliver as much information as we can successfully get from the local cluster state. But definitely not be a regression by delivering incorrect information.